### DISCUSSION TITLE
Save/resume a study doesn't work with a journal backend?

### DISCUSSION THREAD

**Discussion by emnahum** [optuna/optuna#5506] [Q&A]
Hi,  I've been using optuna for about 2 months now, and I'm trying to resume a study with a journal backend:

```
storage_location = "/optuna/optuna.log"
print("optuna: storage location is ", storage_location)

study_name = global_vars.model_date + "-" + global_vars.model_num
print("optuna: study name is ", study_name)

optuna.logging.get_logger("optuna").addHandler(logging.StreamHandler(sys.stdout))
storage = optuna.storages.JournalStorage(
    optuna.storages.JournalFileStorage(storage_location)
    )

study = optuna.create_study(direction="minimize", study_name=study_name, storage=storage, load_if_exists=True)
print("Study is ", study)`
```

The idea is, if I'm not satisfied with the result, I can continue the study from where I left off.  But when I did this, it started from scratch again, as shown by this optimization history plot:

![20240521-190-opt-history](https://github.com/optuna/optuna/assets/25746313/644e263a-88a9-49d1-b8ae-34c102fc2395)

The first curve on the left is the original run, about 96 trials.  The second curve is when I decided to run again for another 48 trials, but it started from scratch again (randomized initial values), despite the `load_if_exists=True` parameter.  I tried a third time with 128 trials, and got a better result, but I had to waste a lot of time (and GPU cycles) re-running the first 96 or so trials.  Here's a timeline plot to illustrate the history:

![20240521-190-plot-timeline](https://github.com/optuna/optuna/assets/25746313/b2dff578-a5eb-4968-8dbe-249c93cc4e5c)

You can see the breaks between the three segments, where I ran a new optuna study.

Here's my parameter selections, if it makes any difference:
```
    # optuna parameters
    batch_size    = int(this_trial.suggest_categorical("batch_size", ["16", "32", "64", "128"]))
    learning_rate = this_trial.suggest_loguniform("learning_rate", 1e-6, 1e-5)
    weight_decay  = this_trial.suggest_loguniform("weight_decay", 1e-7, 1e-5)
```

This is using the HuggingFace trainer.  I don't see why this doesn't work.  My only theory: save/resume doesn't work with a journal backend, but I don't understand why.

Thanks,

-Erich

**Comment by HideakiImamura** [comment]
Thanks for raising the problem. I'm one of maintainers of Optuna. At first, could you tell me the version of Optuna? And, if you have specified the sampler, could you tell me that?

**Comment by emnahum** [comment]
Hey, thanks for responding.

I'm using the latest, 3.6.1.

I have not specified a sampler nor a pruner.  I'm trying to keep it simple
for now.

I'm using HuggingFace, but not the trainer.hyperparameter_search()
function.  Instead, I'm calling trainer.train() inside the
objective_function so that I can save the model if I want.

I'm trying to reproduce it using a simple example from your site (e.g,
https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/001_rdb.html#rdb),
but it's not happening.  My current thought is that a) the code is too
complex (3 factors being evaluated) or b) that it's something to do with
trainer.train.  Those are just speculations, FWIW.

-Erich

**Comment by not522** [comment]
Thanks for your report. I think it is strange that the optimization curve is so nicely decreasing. I suppose that the trial is loading a model that was trained in the previous trial. If it is only re-training from scratch when you resume `Study`, that would explain your report. Could you please check your code?

**Comment by emnahum** [comment]
I'm not sure I understand.  I think it *is* retraining from scratch when I
resume the study.  I would expect that, when I resume the study the next
day, it would get the latest values from trial.suggest_loguniform (for
example), rather than the initial values, which are random numbers.

In terms of code, it is a new python process, but based on the Journal
storage.  When a new run starts, using an old storage, is the suggest()
code not affected?  I.e., the code that suggests() a value is kept in
memory (and thus thrown away at the end of a process), instead of kept in
the storage?

-Erich

**Comment by eukaryo** [comment]
I would like to ask a question about `study_name = global_vars.model_date + "-" + global_vars.model_num` of your code.
Is this `study_name` identical in the three trainings? If not, optimizer does not see the past trials in the studies whose `study_name` is different.

<details>
<summary>Python Code</summary>

```python
import logging
import math
import os
import sys
import time
import uuid

import optuna


def objective(trial):
    time.sleep(1)
    return -trial.number


storage_location = f"./optuna_samplecode_{uuid.uuid4()}.log"
print("optuna: storage location is ", storage_location)


def single_experiment(study_name, n_trials):
    # study_name = global_vars.model_date + "-" + global_vars.model_num
    print("optuna: study name is ", study_name)

    optuna.logging.get_logger("optuna").addHandler(logging.StreamHandler(sys.stdout))
    storage = optuna.storages.JournalStorage(
        optuna.storages.JournalFileStorage(storage_location)
    )

    study = optuna.create_study(
        direction="minimize",
        study_name=study_name,
        storage=storage,
        load_if_exists=True,
    )
    print("Study is ", study)
    study.optimize(objective, n_jobs=12, n_trials=n_trials)


def three_experiments(suffixes):

    if os.path.isfile(storage_location):
        os.remove(storage_location)

    single_experiment(f"study{suffixes[0]}", 96)
    single_experiment(f"study{suffixes[1]}", 48)
    single_experiment(f"study{suffixes[2]}", 128)

    storage = optuna.storages.JournalStorage(
        optuna.storages.JournalFileStorage(storage_location)
    )
    study_for_plot = optuna.create_study()
    for name in optuna.study.get_all_study_names(storage=storage):
        loaded_study = optuna.load_study(study_name=name, storage=storage)
        study_for_plot.add_trials(loaded_study.get_trials())
    fig = optuna.visualization.plot_optimization_history(study_for_plot)
    fig.write_image(f"history{suffixes}.png")


three_experiments("123")
three_experiments("111")

```

</details>
As a result, 

`history123.png` is:
![history123](https://github.com/optuna/optuna/assets/22315555/63a98c09-5fe2-4f8e-8965-94126b76983b)

`history111.png` is:
![history111](https://github.com/optuna/optuna/assets/22315555/66cde9ce-e284-4bef-b258-5877220ce6c9)

**Comment by emnahum** [comment]
Yes, it is the same name in the code.  As I mentioned, I cannot reproduce
it using a simple example like above.

-Erich

**Comment by not522** [comment]
Each trial of Optuna needs to be independent, but I expect that your code does not do so. I think your report can be explained if the model trained in the previous trial is reused in the next trial. For example, it may specify the same directory across trials in the trainer's `resume_from_checkpoint`.

**Comment by emnahum** [comment]
I don't believe that is happening.  While I am using
resume_from_checkpoint=results_dir in trainer.train(), the results_dir is
set to a new directory each trial, based on the values of the
hyperparameters.  Thus, for 224 trials, I have 224 directories.  For
example, with batch size of 128, learning rate of 1.0425895168429575e-06,
and weight decay of 9.830330864534554e-06, I get:

results-regression_nl:1_op:adamw_hf_bs:128_lr:1.0425895168429575e-06_wd:9.830330864534554e-06

Since the values are continuous, the names are unique, and there is
effectively no re-use of directories.

-Erich

**Comment by emnahum** [comment]
BTW, I don't think I'm the only one who has encountered this problem.

From
https://discuss.huggingface.co/t/transformers-and-hyperparameter-search-using-optuna/38429/2
:

b) Optuna’s DB persistence should allow you to interrupt and then resume
the hyperparameters search. While I have that feature work correctly when
using just Optuna, in a toy example, so far I couldn’t make it work with
Hugging Face: when I try to resume the hyperparameters search, it restarts
from the beginning instead, ignoring trials from the previous experiment.
Perhaps I am doing something wrong here.

Perhaps it is a HF bug.

-Erich
